runtime-modules/bounty/src/lib.rs

```diff
@@ -0,0 +1,676 @@
+        /// `O (1)`
+        /// - db:
+        ///    - `O(1)` doesn't depend on the state or parameters
+        /// # </weight>
+        #[weight = WeightInfoBounty::<T>::cancel_bounty_by_member()
+              .max(WeightInfoBounty::<T>::cancel_bounty_by_council())]
+        pub fn cancel_bounty(origin, creator: BountyCreator<MemberId<T>>, bounty_id: T::BountyId) {
+            let bounty_creator_manager = BountyCreatorManager::<T>::get_bounty_creator(
+                origin,
+                creator.clone(),
+            )?;
+
+            ensure!(
+                <Bounties<T>>::contains_key(bounty_id),
+                Error::<T>::BountyDoesntExist
+            );
+
+            let mut bounty = <Bounties<T>>::get(bounty_id);
+
+            bounty_creator_manager.validate_creator(&bounty.creation_params.creator)?;
+
+            let current_bounty_stage = Self::get_bounty_stage(&bounty);
+            ensure!(
+                matches!(current_bounty_stage, BountyStage::Funding),
+                Error::<T>::InvalidBountyStage,
+            );
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            bounty.state = BountyMilestone::Canceled;
+            <Bounties<T>>::insert(bounty_id, bounty);
+
+            Self::deposit_event(RawEvent::BountyCanceled(bounty_id, creator));
+        }
+
+        /// Vetoes a bounty.
+        /// # <weight>
+        ///
+        /// ## weight
+        /// `O (1)`
+        /// - db:
+        ///    - `O(1)` doesn't depend on the state or parameters
+        /// # </weight>
+        #[weight = WeightInfoBounty::<T>::veto_bounty()]
+        pub fn veto_bounty(origin, bounty_id: T::BountyId) {
+            ensure_root(origin)?;
+
+            ensure!(
+                <Bounties<T>>::contains_key(bounty_id),
+                Error::<T>::BountyDoesntExist
+            );
+
+            let mut bounty = <Bounties<T>>::get(bounty_id);
+
+            let current_bounty_stage = Self::get_bounty_stage(&bounty);
+            ensure!(
+                matches!(current_bounty_stage, BountyStage::Funding),
+                Error::<T>::InvalidBountyStage,
+            );
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            bounty.state = BountyMilestone::Canceled;
+            <Bounties<T>>::insert(bounty_id, bounty);
+
+            Self::deposit_event(RawEvent::BountyVetoed(bounty_id));
+        }
+
+        /// Provides bounty funding.
+        /// # <weight>
+        ///
+        /// ## weight
+        /// `O (1)`
+        /// - db:
+        ///    - `O(1)` doesn't depend on the state or parameters
+        /// # </weight>
+        #[weight = WeightInfoBounty::<T>::fund_bounty()]
+        pub fn fund_bounty(
+            origin,
+            member_id: MemberId<T>,
+            bounty_id: T::BountyId,
+            amount: BalanceOf<T>
+        ) {
+            let account_id = T::MemberOriginValidator::ensure_member_controller_account_origin(
```

Comment:
We only allow contributing from controller account? I am fine with that, just confirming that this is what we wanted.

Reply:
Yes. Renamed to `controller_account_id` for clarity.